File: 6_kyu/Find_the_odd_int.c
#include <stddef.h>

// Finds the integer occuring odd times in the array.
// ALL other integers are even times in the array.
// Trick: it is a ^ a = 0 and a ^ a ^ b = b.
int find_odd (const size_t length, const int array[length])
{
  int accu = 0;
  for( size_t i = 0; i < length; ++i )
    accu ^= array[i];
  return accu;
}

File: 6_kyu/Find_the_odd_int.rs
use std::collections::HashSet;

fn find_odd(arr: &[i32]) -> i32 {
    // Set of all integers seen 1, 3, 5, .. times in the array.
    let mut odd_integers = HashSet::new();
    for it in arr {
        if odd_integers.contains(it) {
            odd_integers.remove(it);
        } else {
            odd_integers.insert(it);
        }
    }
    if let Some(result) = odd_integers.iter().next() {
        **result
    } else {
        panic!("No int was odd times in the array.")
    }
}
